### ISSUE TITLE
Assess the correct use of `ConcreteReadAttributePath` in the read call-path for the IM

### ISSUE THREAD

**Issue by mrjerryjohns** [project-chip/connectedhomeip#12049]
This [PR](https://github.com/project-chip/connectedhomeip/pull/11705/files#r753406382) introduces a `ConcreteReadAttributePath` that contains a list index. However, it adds the use of that fairly late in the call chain for handling a read (inside the `ReadSingleClusterData` function). We should assess where the right place for this to be is.

**Comment by mrjerryjohns** [comment]
Assigning this to you @erjiaqing since you're gonna be in this space a fair bit.

**Comment by bzbarsky-apple** [comment]
Fwiw, I think `ReadSingleClusterData` should definitely take a `ConcreteReadAttributePath`.

**Comment by bzbarsky-apple** [comment]
https://github.com/project-chip/connectedhomeip/pull/12067 fixes the obvious thing.  Need to figure out how to share the path-expanding iterator between read and write paths before we can push things further.

**Comment by holbrookt** [comment]
@bzbarsky-apple @mrjerryjohns is this needed for 1.0? If not, please remove the label. Thanks!

**Comment by bzbarsky-apple** [comment]
I don't think the remaining work here is a 1.0 blocker, personally.

**Comment by mrjerryjohns** [comment]
Already done.
